Response: Remove odd response body discard reader

roosterfish opened the pull request:
The ParseResponse function always reads from the response which invalidates the purpose of the body discard reader.
It's good Go practice to always read the response and then close the body but not anymore required in this case.

As a neat side effect we remove the logging statement from the client func which is odd anyway as such an error should rather be returned directly and not logged as this func is not only used as part of the daemon (which has an initialized logger) but can also be called by clients directly.

roosterfish committed:
microcluster/rest/response: Remove odd response body discard reader

The ParseResponse function always reads from the response which invalidates the purpose of the body discard reader.

Signed-off-by: Julian Pelizäus <julian.pelizaeus@canonical.com>

tugbataluy reviewed (dismissed):
Approved. As fat as I understand, this is a focused cleanup that eliminates unnecessary logic, improves architectural separation, and ensures errors are handled properly at the caller level rather than being logged internally.

roosterfish commented:
I'll leave it open as I want to verify it further. Also I want to investigate why this was added in the first place.
As it's not causing any issues rn there is no pressure right now to get it merged. Just something I found along the way.

markylaing commented on microcluster/rest/response/response.go:
Can we also remove this? Or at least allow a range of response codes 200-299? It seems weird that if someone adds a microcluster endpoint like an operation returning a 202 then this function will behave differently in the error scenario

roosterfish replied:
Good point, this entire section was landed in the [initial blob](https://github.com/canonical/microcluster/commit/2f97da8b47340d4e68f22c3b366a9a05742611a0), so I suspect we can try to live without it. Removed.

markylaing reviewed (dismissed):
Good to double check where this came from though I suspect it was just copied from LXD or elsewhere. 

Good to get rid of this cruft though, any errors that result from reading the response body would be caught during the JSON decoding.

roosterfish committed:
microcluster/types: Remove opinionated error check when parsing response

Ensure if the response cannot be parsed we always return the same error unrelated to the response code.

Signed-off-by: Julian Pelizäus <julian.pelizaeus@canonical.com>

markylaing reviewed (approved):
Smashing thanks 👍
